File: src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/WhenKeywordRecommender.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis.Completion.Providers
Imports Microsoft.CodeAnalysis.VisualBasic.Extensions.ContextQuery
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.KeywordRecommenders.Statements
    ''' <summary>
    ''' Recommends the "When" keyword for a Catch filter
    ''' </summary>
    Friend Class WhenKeywordRecommender
        Inherits AbstractKeywordRecommender

        Private Shared ReadOnly s_keywords As ImmutableArray(Of RecommendedKeyword) =
            ImmutableArray.Create(New RecommendedKeyword("When", VBFeaturesResources.Adds_a_conditional_test_to_a_Catch_statement_Exceptions_are_caught_by_that_Catch_statement_only_when_the_conditional_test_that_follows_the_When_keyword_evaluates_to_True))

        Protected Overrides Function RecommendKeywords(context As VisualBasicSyntaxContext, cancellationToken As CancellationToken) As ImmutableArray(Of RecommendedKeyword)
            If context.FollowsEndOfStatement Then
                Return ImmutableArray(Of RecommendedKeyword).Empty
            End If

            Dim targetToken = context.TargetToken

            If targetToken.IsFromIdentifierNode(Of CatchStatementSyntax)(Function(catchStatement) catchStatement.IdentifierName) Then
                Return s_keywords
            End If

            If context.SyntaxTree.IsFollowingCompleteExpression(Of SimpleAsClauseSyntax)(context.Position, context.TargetToken,
                childGetter:=Function(asClause) If(TypeOf asClause.Parent Is CatchStatementSyntax, asClause.Type, Nothing), cancellationToken:=cancellationToken) Then

                Return s_keywords
            End If

            Return ImmutableArray(Of RecommendedKeyword).Empty
        End Function
    End Class
End Namespace

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Shared.Extensions;

namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders;

internal sealed class WhenKeywordRecommender : AbstractSyntacticSingleKeywordRecommender
{
    public WhenKeywordRecommender()
        : base(SyntaxKind.WhenKeyword, isValidInPreprocessorContext: true)
    {
    }

    protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
    {
        return context.IsCatchFilterContext ||
            IsAfterCompleteExpressionOrPatternInCaseLabel(context) ||
            IsAtEndOfPatternInSwitchExpression(context);
    }

    private static bool IsAtEndOfPatternInSwitchExpression(CSharpSyntaxContext context)
    {
        if (!context.IsAtEndOfPattern)
            return false;

        // `x switch { SomePattern $$
        var pattern = context.TargetToken.GetAncestors<PatternSyntax>().LastOrDefault();
        if (pattern?.Parent is SwitchExpressionArmSyntax)
            return true;

        return false;
    }

    private static bool IsAfterCompleteExpressionOrPatternInCaseLabel(CSharpSyntaxContext context)
    {
        var switchLabel = context.TargetToken.GetAncestor<SwitchLabelSyntax>();
        if (switchLabel == null)
            return false;

        var expressionOrPattern = switchLabel.ChildNodes().FirstOrDefault();
        if (expressionOrPattern == null)
        {
            // It must have been a default label.
            return false;
        }

        // Never show `when` after `var` in a pattern.  It's virtually always going to be unhelpful as the user is
        // far more likely to be writing `case var someName...` rather than typing `cae var when...` (in the case
        // that `var` is a constant).  In other words, it's fine to make that rare case have to manually type out
        // `when` rather than risk having `when` pop up when it's not desired.
        if (context.TargetToken.Text == "var")
            return false;

        // If the last token is missing, the expression is incomplete - possibly because of missing parentheses,
        // but not necessarily. We don't want to offer 'when' in those cases. Here are some examples that illustrate this:
        // case |
        // case x.|
        // case 1 + |
        // case (1 + 1 |

        // Also note that if there's a missing token inside the expression, that's fine and we do offer 'when':
        // case (1 + ) |

        if (expressionOrPattern.GetLastToken(includeZeroWidth: true).IsMissing)
            return false;

        var lastToken = expressionOrPattern.GetLastToken(includeZeroWidth: false);

        // We're writing past the end of a complete pattern.  This is a place where 'when' could be added to add
        // restrictions on the pattern.
        if (lastToken == context.TargetToken)
            return true;

        // We're writing the last token of a pattern.  In this case, we might either be writing a name for the pattern
        // (like `case Wolf w:`) or we might be starting to write `when` (like `case Wolf when ...:`).
        if (lastToken == context.LeftToken)
        {
            if (expressionOrPattern is DeclarationPatternSyntax)
                return true;

            if (expressionOrPattern is RecursivePatternSyntax)
                return true;
        }

        return false;
    }
}
